File: backend/tests/api.integration.test.ts
import { describe, it, expect, beforeAll } from "vitest";

const BASE = process.env.API_BASE_URL || "http://127.0.0.1:4000";

type LoginRes = {
  token: string;
  user: { id: string; email: string; username: string; role: string };
};

async function api(path: string, init: RequestInit = {}) {
  const res = await fetch(`${BASE}${path}`, {
    ...init,
    headers: {
      Accept: "application/json",
      ...(init.body ? { "Content-Type": "application/json" } : {}),
      ...(init.headers || {}),
    },
  });
  const text = await res.text();
  let json: unknown = null;
  try {
    json = text ? JSON.parse(text) : null;
  } catch {
    json = text;
  }
  return { res, json, status: res.status };
}

async function login(email: string, password: string): Promise<LoginRes> {
  const { res, json, status } = await api("/api/auth/login", {
    method: "POST",
    body: JSON.stringify({ email, password }),
  });
  expect(status).toBe(200);
  expect(res.ok).toBe(true);
  return json as LoginRes;
}

describe("API — Health", () => {
  it("TC-API-01: GET /api/health zwraca status ok i db up", async () => {
    const { status, json } = await api("/api/health");
    expect(status).toBe(200);
    expect(json).toMatchObject({ status: "ok", db: "up", service: "raceportal-api" });
  });
});

describe("API — Auth", () => {
  it("TC-API-02: logowanie kierowcy (seed) zwraca JWT i rolę USER", async () => {
    const data = await login("test@wp.pl", "test123");
    expect(data.token).toBeTruthy();
    expect(data.user.role).toBe("USER");
    expect(data.user.email).toBe("test@wp.pl");
  });

  it("TC-API-03: logowanie admina zwraca rolę ADMIN", async () => {
    const data = await login("admin@raceportal.pl", "admin123");
    expect(data.user.role).toBe("ADMIN");
  });

  it("TC-API-04: logowanie organizatora zwraca rolę ORGANIZER", async () => {
    const data = await login("org@raceportal.pl", "org123");
    expect(data.user.role).toBe("ORGANIZER");
  });

  it("TC-API-05: błędne hasło → 401", async () => {
    const { status, json } = await api("/api/auth/login", {
      method: "POST",
      body: JSON.stringify({ email: "test@wp.pl", password: "zle-haslo" }),
    });
    expect(status).toBe(401);
    expect((json as { error: string }).error).toBeTruthy();
  });

  it("TC-API-06: /api/auth/me bez tokena → 401", async () => {
    const { status } = await api("/api/auth/me");
    expect(status).toBe(401);
  });

  it("TC-API-07: /api/auth/me z tokenem zwraca profil", async () => {
    const { token } = await login("test@wp.pl", "test123");
    const { status, json } = await api("/api/auth/me", {
      headers: { Authorization: `Bearer ${token}` },
    });
    expect(status).toBe(200);
    expect((json as { email: string }).email).toBe("test@wp.pl");
  });

  it("TC-API-08: rejestracja nowego użytkownika", async () => {
    const email = `test.auto.${Date.now()}@raceportal.test`;
    const { status, json } = await api("/api/auth/register", {
      method: "POST",
      body: JSON.stringify({
        username: "AutoTester",
        email,
        password: "test1234",
      }),
    });
    expect(status).toBe(201);
    const data = json as LoginRes;
    expect(data.token).toBeTruthy();
    expect(data.user.email).toBe(email);
    expect(data.user.role).toBe("USER");
  });
});

describe("API — Events", () => {
  it("TC-API-09: lista wydarzeń jest paginowana i zawiera APPROVED przyszłe", async () => {
    const { status, json } = await api("/api/events?limit=12");
    expect(status).toBe(200);
    const data = json as { items: unknown[]; total: number; page: number; limit: number };
    expect(data.page).toBe(1);
    expect(data.limit).toBe(12);
    expect(Array.isArray(data.items)).toBe(true);
    expect(data.total).toBeGreaterThanOrEqual(1);
    expect(data.items.length).toBeGreaterThanOrEqual(1);
  });

  it("TC-API-10: filtr wyszukiwania q zwraca dopasowania", async () => {
    const { status, json } = await api("/api/events?q=Pozna%C5%84&limit=20");
    expect(status).toBe(200);
    const data = json as { items: Array<{ name: string; city: string; track: string }> };
    expect(data.items.length).toBeGreaterThanOrEqual(1);
    const hit = data.items.some(
      (e) =>
        e.city.toLowerCase().includes("poznań") ||
        e.track.toLowerCase().includes("poznań") ||
        e.name.toLowerCase().includes("poznań") ||
        e.name.toLowerCase().includes("mistrzostwa"),
    );
    expect(hit).toBe(true);
  });

  it("TC-API-11: szczegóły wydarzenia po id", async () => {
    const list = await api("/api/events?limit=1");
    const id = (list.json as { items: Array<{ id: string }> }).items[0].id;
    const { status, json } = await api(`/api/events/${id}`);
    expect(status).toBe(200);
    expect((json as { id: string }).id).toBe(id);
    expect((json as { description: string }).description.length).toBeGreaterThan(0);
  });

  it("TC-API-12: archiwum zwraca wydarzenia historyczne", async () => {
    const { status, json } = await api("/api/events?archive=1&limit=20");
    expect(status).toBe(200);
    const data = json as { items: Array<{ status: string }>; total: number };
    expect(data.total).toBeGreaterThanOrEqual(1);
    expect(data.items.every((e) => e.status === "ARCHIVED" || e.status === "APPROVED")).toBe(true);
  });

  it("TC-API-13: kategorie meta", async () => {
    const { status, json } = await api("/api/events/meta/categories");
    expect(status).toBe(200);
    expect(Array.isArray(json)).toBe(true);
    expect((json as string[]).length).toBeGreaterThanOrEqual(1);
  });
});

describe("API — Garage & Registrations", () => {
  let driverToken: string;
  let eventId: string;

  beforeAll(async () => {
    driverToken = (await login("test@wp.pl", "test123")).token;
    const list = await api("/api/events?limit=1");
    eventId = (list.json as { items: Array<{ id: string }> }).items[0].id;
  });

  it("TC-API-14: garaż kierowcy zwraca listę aut", async () => {
    const { status, json } = await api("/api/garage", {
      headers: { Authorization: `Bearer ${driverToken}` },
    });
    expect(status).toBe(200);
    expect(Array.isArray(json)).toBe(true);
    expect((json as unknown[]).length).toBeGreaterThanOrEqual(1);
  });

  it("TC-API-15: zgłoszenie na wydarzenie", async () => {
    const { status, json } = await api("/api/registrations", {
      method: "POST",
      headers: { Authorization: `Bearer ${driverToken}` },
      body: JSON.stringify({ eventId, note: "Test automatyczny" }),
    });
    expect([200, 201]).toContain(status);
    expect((json as { eventId: string }).eventId).toBe(eventId);
  });

  it("TC-API-16: lista moich zgłoszeń", async () => {
    const { status, json } = await api("/api/registrations/mine", {
      headers: { Authorization: `Bearer ${driverToken}` },
    });
    expect(status).toBe(200);
    expect(Array.isArray(json)).toBe(true);
  });
});

describe("API — RBAC Admin / Organizer", () => {
  it("TC-API-17: panel admin stats wymaga ADMIN", async () => {
    const user = await login("test@wp.pl", "test123");
    const denied = await api("/api/admin/stats", {
      headers: { Authorization: `Bearer ${user.token}` },
    });
    expect(denied.status).toBe(403);

    const admin = await login("admin@raceportal.pl", "admin123");
    const ok = await api("/api/admin/stats", {
      headers: { Authorization: `Bearer ${admin.token}` },
    });
    expect(ok.status).toBe(200);
    expect(ok.json).toHaveProperty("users");
    expect(ok.json).toHaveProperty("events");
  });

  it("TC-API-18: organizer events dostępne dla ORGANIZER", async () => {
    const org = await login("org@raceportal.pl", "org123");
    const { status, json } = await api("/api/organizer/events", {
      headers: { Authorization: `Bearer ${org.token}` },
    });
    expect(status).toBe(200);
    expect(Array.isArray(json)).toBe(true);
  });

  it("TC-API-19: pending events widoczne dla admina", async () => {
    const admin = await login("admin@raceportal.pl", "admin123");
    const { status, json } = await api("/api/admin/events/pending", {
      headers: { Authorization: `Bearer ${admin.token}` },
    });
    expect(status).toBe(200);
    expect(Array.isArray(json)).toBe(true);
  });
});

describe("API — Maps", () => {
  it("TC-API-20: wytyczanie trasy (OSRM/Google)", async () => {
    const { status, json } = await api("/api/maps/route", {
      method: "POST",
      body: JSON.stringify({
        fromLat: 52.23,
        fromLng: 21.01,
        toLat: 52.33,
        toLng: 16.85,
      }),
    });
    expect(status).toBe(200);
    const data = json as {
      provider: string;
      distanceText: string;
      durationText: string;
      polyline: unknown[];
    };
    expect(["osrm", "google"]).toContain(data.provider);
    expect(data.distanceText).toBeTruthy();
    expect(data.polyline.length).toBeGreaterThan(1);
  });
});

File: backend/src/test/java/pl/raceportal/ApiIntegrationTest.java
package pl.raceportal;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.condition.EnabledIf;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;
import org.springframework.test.web.servlet.MockMvc;
import org.testcontainers.DockerClientFactory;
import org.testcontainers.containers.MySQLContainer;
import pl.raceportal.domain.Event;
import pl.raceportal.domain.EventStatus;
import pl.raceportal.repository.EventRepository;

import java.math.BigDecimal;

import static org.hamcrest.Matchers.contains;
import static org.hamcrest.Matchers.containsString;
import static org.hamcrest.Matchers.greaterThanOrEqualTo;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.notNullValue;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.patch;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

/**
 * Testy integracyjne API RacePortal na prawdziwej bazie MySQL.
 * <p>
 * Rola w architekturze testów: end-to-end przez MockMvc — auth, wydarzenia, zgłoszenia,
 * garaż, RBAC. Preferuje Testcontainers ({@code mysql:8.0}); alternatywnie
 * zewnętrzne {@code TEST_DB_URL} / {@code TEST_DB_USER} / {@code TEST_DB_PASSWORD}.
 * Pomijane gdy brak Dockera i braku zewnętrznej DB ({@code @EnabledIf}).
 * </p>
 * Technologie: Spring Boot Test, MockMvc, Testcontainers MySQL, JUnit 5, profil {@code test}.
 * <p>
 * Pomysł (alt): RestAssured; Testcontainers + shared container (Ryuk); WireMock dla OSRM/Mail.
 * </p>
 */
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.MOCK)
@AutoConfigureMockMvc
@ActiveProfiles("test")
@EnabledIf("pl.raceportal.ApiIntegrationTest#dockerOrExternalDbAvailable")
class ApiIntegrationTest {

    /** Czy używać zewnętrznej DB zamiast Testcontainers. */
    private static final boolean USE_EXTERNAL_DB = System.getenv("TEST_DB_URL") != null
            && !System.getenv("TEST_DB_URL").isBlank();

    private static MySQLContainer<?> mysql;

    static {
        if (!USE_EXTERNAL_DB && DockerClientFactory.instance().isDockerAvailable()) {
            mysql = new MySQLContainer<>("mysql:8.0")
                    .withDatabaseName("raceportal")
                    .withUsername("raceportal")
                    .withPassword("raceportal");
            mysql.start();
        }
    }

    /** Warunek uruchomienia klasy — Docker lub skonfigurowana zewnętrzna baza. */
    @SuppressWarnings("unused")
    static boolean dockerOrExternalDbAvailable() {
        return USE_EXTERNAL_DB || DockerClientFactory.instance().isDockerAvailable();
    }

    /** Podłącza datasource Spring do kontenera / env vars. */
    @DynamicPropertySource
    static void datasourceProperties(DynamicPropertyRegistry registry) {
        if (USE_EXTERNAL_DB) {
            registry.add("spring.datasource.url", () -> System.getenv("TEST_DB_URL"));
            registry.add("spring.datasource.username",
                    () -> System.getenv().getOrDefault("TEST_DB_USER", "raceportal"));
            registry.add("spring.datasource.password",
                    () -> System.getenv().getOrDefault("TEST_DB_PASSWORD", "raceportal"));
        } else if (mysql != null) {
            registry.add("spring.datasource.url", mysql::getJdbcUrl);
            registry.add("spring.datasource.username", mysql::getUsername);
            registry.add("spring.datasource.password", mysql::getPassword);
        }
    }

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private ObjectMapper objectMapper;

    @Autowired
    private EventRepository eventRepository;

    private String login(String email, String password) throws Exception {
        String body = objectMapper.writeValueAsString(new LoginPayload(email, password));
        String response = mockMvc.perform(post("/api/auth/login")
                        .contentType("application/json")
                        .content(body))
                .andExpect(status().isOk())
                .andReturn().getResponse().getContentAsString();
        JsonNode node = objectMapper.readTree(response);
        return node.get("token").asText();
    }

    /** Healthcheck publiczny — status ok + db up. */
    @Test
    void health_returnsOkWithDatabaseUp() throws Exception {
        mockMvc.perform(get("/api/health"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.status", is("ok")))
                .andExpect(jsonPath("$.service", is("raceportal-api")))
                .andExpect(jsonPath("$.db", is("up")));
    }

    /** Seed: admin, organizer i kierowca mogą się zalogować. */
    @Test
    void login_worksForAllThreeSeededRoles() throws Exception {
        String adminToken = login("admin@raceportal.pl", "admin123");
        String orgToken = login("org@raceportal.pl", "org123");
        String userToken = login("test@wp.pl", "test123");

        assertNotBlank(adminToken);
        assertNotBlank(orgToken);
        assertNotBlank(userToken);
    }

    @Test
    void login_withWrongPassword_returns401() throws Exception {
        String body = objectMapper.writeValueAsString(new LoginPayload("test@wp.pl", "wrong-password"));
        mockMvc.perform(post("/api/auth/login")
                        .contentType("application/json")
                        .content(body))
                .andExpect(status().isUnauthorized())
                .andExpect(jsonPath("$.error", notNullValue()));
    }

    @Test
    void me_withoutToken_returns401() throws Exception {
        mockMvc.perform(get("/api/auth/me"))
                .andExpect(status().isUnauthorized())
                .andExpect(jsonPath("$.error", notNullValue()));
    }

    @Test
    void me_withValidToken_returns200() throws Exception {
        String token = login("test@wp.pl", "test123");
        mockMvc.perform(get("/api/auth/me").header("Authorization", "Bearer " + token))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.email", is("test@wp.pl")))
                .andExpect(jsonPath("$.role", is("USER")));
    }

    @Test
    void register_withExistingEmail_returns409() throws Exception {
        String body = objectMapper.writeValueAsString(new RegisterPayload("Another", "test@wp.pl", "password123"));
        mockMvc.perform(post("/api/auth/register")
                        .contentType("application/json")
                        .content(body))
                .andExpect(status().isConflict())
                .andExpect(jsonPath("$.error", notNullValue()));
    }

    @Test
    void register_withInvalidPayload_returns400() throws Exception {
        String body = objectMapper.writeValueAsString(new RegisterPayload("A", "not-an-email", "123"));
        mockMvc.perform(post("/api/auth/register")
                        .contentType("application/json")
                        .content(body))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.error", notNullValue()))
                .andExpect(jsonPath("$.details", notNullValue()));
    }

    /** Flow rejestracji lokalnej: register → verify-email → JWT. */
    @Test
    void register_thenVerifyEmail_returnsToken() throws Exception {
        String email = "nowy.kierowca@example.com";
        String registerBody = objectMapper.writeValueAsString(new RegisterPayload("NowyKierowca", email, "password123"));
        mockMvc.perform(post("/api/auth/register")
                        .contentType("application/json")
                        .content(registerBody))
                .andExpect(status().isCreated())
                .andExpect(jsonPath("$.requiresVerification", is(true)))
                .andExpect(jsonPath("$.email", is(email)));

        String loginAttemptBody = objectMapper.writeValueAsString(new LoginPayload(email, "password123"));
        mockMvc.perform(post("/api/auth/login")
                        .contentType("application/json")
                        .content(loginAttemptBody))
                .andExpect(status().isForbidden());

        String wrongCodeBody = objectMapper.writeValueAsString(new VerifyEmailPayload(email, "000000"));
        mockMvc.perform(post("/api/auth/verify-email")
                        .contentType("application/json")
                        .content(wrongCodeBody))
                .andExpect(status().isBadRequest());
    }

    @Test
    void events_list_returnsSeededApprovedEvents() throws Exception {
        mockMvc.perform(get("/api/events").param("page", "1").param("limit", "12"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.items").isArray())
                .andExpect(jsonPath("$.total", greaterThanOrEqualTo(1)));
    }

    @Test
    void garage_list_returnsSeededCarsForDriver() throws Exception {
        String token = login("test@wp.pl", "test123");
        mockMvc.perform(get("/api/garage").header("Authorization", "Bearer " + token))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$", org.hamcrest.Matchers.hasSize(org.hamcrest.Matchers.greaterThanOrEqualTo(2))));
    }

    /**
     * Reguła garażu: przy otwartym zgłoszeniu nie wolno zmienić make/model/class,
     * ale inne pola (np. plate) są dozwolone.
     */
    @Test
    void garage_update_blockedByOpenRegistration_thenAllowedForNonConflictingFields() throws Exception {
        String token = login("test@wp.pl", "test123");

        String carBody = objectMapper.writeValueAsString(
                new CarPayload("Toyota", "GR86", 2023, null, "PO 11111", null));
        String carResponse = mockMvc.perform(post("/api/garage")
                        .header("Authorization", "Bearer " + token)
                        .contentType("application/json")
                        .content(carBody))
                .andExpect(status().isCreated())
                .andReturn().getResponse().getContentAsString();
        String carId = objectMapper.readTree(carResponse).get("id").asText();

        Event freeEvent = eventRepository.findAll().stream()
                .filter(e -> !e.isPaid() && e.getStatus() == EventStatus.APPROVED)
                .findFirst().orElseThrow();

        String regBody = objectMapper.writeValueAsString(
                new RegistrationCreatePayload(freeEvent.getId(), carId, null));
        mockMvc.perform(post("/api/registrations")
                        .header("Authorization", "Bearer " + token)
                        .contentType("application/json")
                        .content(regBody))
                .andExpect(status().isCreated());

        String conflictingUpdate = objectMapper.writeValueAsString(
                new CarPayload(null, "GR86 Facelift", null, null, null, null));
        mockMvc.perform(patch("/api/garage/" + carId)
                        .header("Authorization", "Bearer " + token)
                        .contentType("application/json")
                        .content(conflictingUpdate))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.error", containsString("otwarte zgłoszenie")));

        mockMvc.perform(delete("/api/garage/" + carId)
                        .header("Authorization", "Bearer " + token))
                .andExpect(status().isBadRequest());

        String plateOnlyUpdate = objectMapper.writeValueAsString(
                new CarPayload(null, null, null, null, "PO 22222", null));
        mockMvc.perform(patch("/api/garage/" + carId)
                        .header("Authorization", "Bearer " + token)
                        .contentType("application/json")
                        .content(plateOnlyUpdate))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.plate", is("PO 22222")));
    }

    /** Anulowanie zgłoszenia przez kierowcę → status CANCELED. */
    @Test
    void driver_cancelRegistration_setsCanceledStatus() throws Exception {
        String token = login("test@wp.pl", "test123");

        String carBody = objectMapper.writeValueAsString(
                new CarPayload("Honda", "Civic Type R", 2020, "Track Day", "GD 33333", null));
        String carResponse = mockMvc.perform(post("/api/garage")
                        .header("Authorization", "Bearer " + token)
                        .contentType("application/json")
                        .content(carBody))
                .andExpect(status().isCreated())
                .andReturn().getResponse().getContentAsString();
        String carId = objectMapper.readTree(carResponse).get("id").asText();

        Event trackDayEvent = eventRepository.findAll().stream()
                .filter(e -> !e.isPaid() && e.getStatus() == EventStatus.APPROVED
                        && "Track Day".equalsIgnoreCase(e.getCategory()))
                .findFirst().orElseThrow();

        String regBody = objectMapper.writeValueAsString(
                new RegistrationCreatePayload(trackDayEvent.getId(), carId, null));
        String regResponse = mockMvc.perform(post("/api/registrations")
                        .header("Authorization", "Bearer " + token)
                        .contentType("application/json")
                        .content(regBody))
                .andExpect(status().isCreated())
                .andReturn().getResponse().getContentAsString();
        String regId = objectMapper.readTree(regResponse).get("id").asText();

        mockMvc.perform(post("/api/registrations/" + regId + "/cancel")
                        .header("Authorization", "Bearer " + token))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.status", is("CANCELED")));
    }

    /**
     * Flow płatny: PENDING → ACCEPTED (org) → payment-proof (kierowca) → CONFIRMED (org).
     */
    @Test
    void paidRegistration_acceptedThenPaymentProofThenConfirmed() throws Exception {
        String driverToken = login("test@wp.pl", "test123");
        String orgToken = login("org@raceportal.pl", "org123");

        Event paidEvent = eventRepository.findAll().stream()
                .filter(Event::isPaid)
                .findFirst().orElseThrow();

        String regBody = objectMapper.writeValueAsString(
                new RegistrationCreatePayload(paidEvent.getId(), null, null));
        String regResponse = mockMvc.perform(post("/api/registrations")
                        .header("Authorization", "Bearer " + driverToken)
                        .contentType("application/json")
                        .content(regBody))
                .andExpect(status().isCreated())
                .andReturn().getResponse().getContentAsString();
        String regId = objectMapper.readTree(regResponse).get("id").asText();

        String acceptBody = objectMapper.writeValueAsString(
                new RegistrationStatusPayload("ACCEPTED", "Prosimy o wpłatę wpisowego"));
        mockMvc.perform(patch("/api/registrations/" + regId + "/status")
                        .header("Authorization", "Bearer " + orgToken)
                        .contentType("application/json")
                        .content(acceptBody))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.status", is("ACCEPTED")))
                .andExpect(jsonPath("$.paymentDueAt", notNullValue()));

        String proofBody = objectMapper.writeValueAsString(new PaymentProofPayload("https://example.com/proof.png"));
        mockMvc.perform(post("/api/registrations/" + regId + "/payment-proof")
                        .header("Authorization", "Bearer " + driverToken)
                        .contentType("application/json")
                        .content(proofBody))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.paymentProofUrl", is("https://example.com/proof.png")));

        String confirmBody = objectMapper.writeValueAsString(new RegistrationStatusPayload("CONFIRMED", null));
        mockMvc.perform(patch("/api/registrations/" + regId + "/status")
                        .header("Authorization", "Bearer " + orgToken)
                        .contentType("application/json")
                        .content(confirmBody))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.status", is("CONFIRMED")));
    }

    /** Alias kategorii: auto KJS powinno pasować do eventu Rally. */
    @Test
    void registration_acceptsCategoryAliasBetweenCarAndEvent() throws Exception {
        String orgToken = login("org@raceportal.pl", "org123");
        String adminToken = login("admin@raceportal.pl", "admin123");
        String driverToken = login("test@wp.pl", "test123");

        String eventBody = objectMapper.writeValueAsString(new EventCreatePayload(
                "Rally alias test",
                "Event testowy dla dopasowania aliasów kategorii.",
                "Rally", "2026-12-15", "10:00", null, null, "Tor Alias", "Aliasowo", "Mazowieckie",
                null, null, null, false, null, null, null, null, true,
                null, null, null, null, null, null));
        String eventResponse = mockMvc.perform(post("/api/events")
                        .header("Authorization", "Bearer " + orgToken)
                        .contentType("application/json")
                        .content(eventBody))
                .andExpect(status().isCreated())
                .andReturn().getResponse().getContentAsString();
        String eventId = objectMapper.readTree(eventResponse).get("id").asText();

        String approveBody = objectMapper.writeValueAsString(new EventStatusPayload("APPROVED"));
        mockMvc.perform(patch("/api/admin/events/" + eventId + "/status")
                        .header("Authorization", "Bearer " + adminToken)
                        .contentType("application/json")
                        .content(approveBody))
                .andExpect(status().isOk());

        String carBody = objectMapper.writeValueAsString(
                new CarPayload("Subaru", "Impreza", 2018, "KJS", "WA 12345", null));
        String carResponse = mockMvc.perform(post("/api/garage")
                        .header("Authorization", "Bearer " + driverToken)
                        .contentType("application/json")
                        .content(carBody))
                .andExpect(status().isCreated())
                .andReturn().getResponse().getContentAsString();
        String carId = objectMapper.readTree(carResponse).get("id").asText();

        String regBody = objectMapper.writeValueAsString(
                new RegistrationCreatePayload(eventId, carId, null));
        mockMvc.perform(post("/api/registrations")
                        .header("Authorization", "Bearer " + driverToken)
                        .contentType("application/json")
                        .content(regBody))
                .andExpect(status().isCreated())
                .andExpect(jsonPath("$.status", is("PENDING")));
    }

    /** Wymóg OC: event z requireOc=true odrzuca auto bez OC. */
    @Test
    void registration_rejectsCarWithoutOcWhenEventRequiresOc() throws Exception {
        String orgToken = login("org@raceportal.pl", "org123");
        String adminToken = login("admin@raceportal.pl", "admin123");
        String driverToken = login("test@wp.pl", "test123");

        String eventBody = objectMapper.writeValueAsString(new EventCreatePayload(
                "OC required test",
                "Event testowy dla walidacji wymogu OC.",
                "Track Day", "2026-12-20", "09:00", null, null, "Tor OC", "Poznań", "Wielkopolskie",
                null, null, null, false, null, null, null, null, true,
                null, null, true, null, null, null));
        String eventResponse = mockMvc.perform(post("/api/events")
                        .header("Authorization", "Bearer " + orgToken)
                        .contentType("application/json")
                        .content(eventBody))
                .andExpect(status().isCreated())
                .andReturn().getResponse().getContentAsString();
        String eventId = objectMapper.readTree(eventResponse).get("id").asText();

        String approveBody = objectMapper.writeValueAsString(new EventStatusPayload("APPROVED"));
        mockMvc.perform(patch("/api/admin/events/" + eventId + "/status")
                        .header("Authorization", "Bearer " + adminToken)
                        .contentType("application/json")
                        .content(approveBody))
                .andExpect(status().isOk());

        String carBody = objectMapper.writeValueAsString(
                new CarPayload("Honda", "Civic", 2017, "Track Day", "PO 44556", null));
        String carResponse = mockMvc.perform(post("/api/garage")
                        .header("Authorization", "Bearer " + driverToken)
                        .contentType("application/json")
                        .content(carBody))
                .andExpect(status().isCreated())
                .andReturn().getResponse().getContentAsString();
        String carId = objectMapper.readTree(carResponse).get("id").asText();

        String regBody = objectMapper.writeValueAsString(
                new RegistrationCreatePayload(eventId, carId, null));
        mockMvc.perform(post("/api/registrations")
                        .header("Authorization", "Bearer " + driverToken)
                        .contentType("application/json")
                        .content(regBody))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.error", containsString("OC")));
    }

    /** Anulowanie wydarzenia kaskadowo ustawia otwarte zgłoszenia na CANCELED. */
    @Test
    void event_cancel_cancelsOpenRegistrations() throws Exception {
        String orgToken = login("org@raceportal.pl", "org123");
        String adminToken = login("admin@raceportal.pl", "admin123");
        String driverToken = login("test@wp.pl", "test123");

        String eventBody = objectMapper.writeValueAsString(new EventCreatePayload(
                "Testowy Trackday do anulowania",
                "Wydarzenie testowe tworzone przez test integracyjny w celu weryfikacji anulowania.",
                "Track Day", "2026-12-01", "10:00", null, null, "Tor Testowy", "Testowo", "Testowe",
                null, null, null, false, null, null, null, null, true,
                null, null, null, null, null, null));
        String eventResponse = mockMvc.perform(post("/api/events")
                        .header("Authorization", "Bearer " + orgToken)
                        .contentType("application/json")
                        .content(eventBody))
                .andExpect(status().isCreated())
                .andReturn().getResponse().getContentAsString();
        String eventId = objectMapper.readTree(eventResponse).get("id").asText();

        String approveBody = objectMapper.writeValueAsString(new EventStatusPayload("APPROVED"));
        mockMvc.perform(patch("/api/admin/events/" + eventId + "/status")
                        .header("Authorization", "Bearer " + adminToken)
                        .contentType("application/json")
                        .content(approveBody))
                .andExpect(status().isOk());

        String regBody = objectMapper.writeValueAsString(new RegistrationCreatePayload(eventId, null, null));
        String regResponse = mockMvc.perform(post("/api/registrations")
                        .header("Authorization", "Bearer " + driverToken)
                        .contentType("application/json")
                        .content(regBody))
                .andExpect(status().isCreated())
                .andReturn().getResponse().getContentAsString();
        String regId = objectMapper.readTree(regResponse).get("id").asText();

        mockMvc.perform(post("/api/events/" + eventId + "/cancel")
                        .header("Authorization", "Bearer " + orgToken))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.status", is("CANCELLED")));

        mockMvc.perform(get("/api/registrations/mine")
                        .header("Authorization", "Bearer " + driverToken))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$[?(@.id == '" + regId + "')].status", contains("CANCELED")));
    }

    /** RBAC: zwykły USER nie wchodzi na /api/admin. */
    @Test
    void admin_endpoint_forbiddenForRegularUser() throws Exception {
        String token = login("test@wp.pl", "test123");
        mockMvc.perform(get("/api/admin/stats").header("Authorization", "Bearer " + token))
                .andExpect(status().isForbidden())
                .andExpect(jsonPath("$.error", notNullValue()));
    }

    /** RBAC: ADMIN ma dostęp do /api/admin/stats. */
    @Test
    void admin_endpoint_allowedForAdmin() throws Exception {
        String token = login("admin@raceportal.pl", "admin123");
        mockMvc.perform(get("/api/admin/stats").header("Authorization", "Bearer " + token))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.users", greaterThanOrEqualTo(1)));
    }

    private static void assertNotBlank(String value) {
        if (value == null || value.isBlank()) {
            throw new AssertionError("Expected non-blank value");
        }
    }

    private record LoginPayload(String email, String password) {
    }

    private record RegisterPayload(String username, String email, String password) {
    }

    private record VerifyEmailPayload(String email, String code) {
    }

    private record CarPayload(String make, String model, Integer year, String className, String plate, String imageUrl) {
    }

    private record RegistrationCreatePayload(String eventId, String carId, String note) {
    }

    private record RegistrationStatusPayload(String status, String comment) {
    }

    private record PaymentProofPayload(String paymentProofUrl) {
    }

    private record EventStatusPayload(String status) {
    }

    private record EventCreatePayload(String name, String description, String category, String date, String time,
                                      String endDate, String endTime, String track, String city, String voivodeship,
                                      String imageUrl, Double lat, Double lng, boolean paid, BigDecimal entryFee,
                                      String bankAccount, Integer paymentDeadlineHours, Integer freeCancelDays,
                                      Boolean acceptRegistrations, BigDecimal spectatorFee, String externalUrl,
                                      Boolean requireOc, Boolean requirePt, Boolean requireCage,
                                      Boolean requireRegistered) {
    }
}
